crates/asm/txs/bridge-v1/src/unstake/script.rs

```diff
@@ -0,0 +1,257 @@
+use bitcoin::{
+    ScriptBuf, XOnlyPublicKey,
+    opcodes::all::{OP_CHECKSIGVERIFY, OP_EQUAL, OP_EQUALVERIFY, OP_SHA256, OP_SIZE},
+    script::Instruction,
+};
+
+/// Instruction indices for the stake connector script
+const PUBKEY_INDEX: usize = 0;
+const STAKE_HASH_INDEX: usize = 6;
+
+/// Builds the stake connector script used in unstaking transactions.
+///
+/// This script validates:
+/// - A signature from the N/N aggregated key
+/// - A 32-byte preimage whose SHA256 hash matches the provided stake_hash
+///
+/// This function serves dual purposes:
+/// 1. Building scripts for new transactions
+/// 2. Validating parsed scripts via reconstruction and comparison
+pub fn stake_connector_script(stake_hash: [u8; 32], pubkey: XOnlyPublicKey) -> ScriptBuf {
+    ScriptBuf::builder()
+        // Verify the signature
+        .push_slice(pubkey.serialize())
+        .push_opcode(OP_CHECKSIGVERIFY)
+        // Verify size of preimage is 32 bytes
+        .push_opcode(OP_SIZE)
+        .push_int(0x20)
+        .push_opcode(OP_EQUALVERIFY)
+        // Verify the preimage matches the hash
+        .push_opcode(OP_SHA256)
+        .push_slice(stake_hash)
+        .push_opcode(OP_EQUAL)
+        .into_script()
+}
+
+/// Extracts the two dynamic parameters (pubkey and stake_hash) from a stake connector script.
+///
+/// This is a minimal extraction that only validates the 32-byte push instructions exist
+/// at the expected positions. Full structural validation happens by reconstructing
+/// the script and comparing byte-for-byte.
+///
+/// Returns `None` if the basic structure doesn't allow parameter extraction.
+fn extract_script_params(script: &ScriptBuf) -> Option<([u8; 32], [u8; 32])> {
+    // Extract pubkey from instruction at PUBKEY_INDEX
+    let pubkey = match script.instructions().nth(PUBKEY_INDEX) {
+        Some(Ok(Instruction::PushBytes(bytes))) if bytes.len() == 32 => {
+            bytes.as_bytes().try_into().ok()?
+        }
+        _ => return None,
+    };
+
+    // Extract stake_hash from instruction at STAKE_HASH_INDEX
+    let stake_hash = match script.instructions().nth(STAKE_HASH_INDEX) {
```

Comment:
This feels like something the compiler would optimize anyway but you shouldn't have to create the iterator twice. Either way this is too tiny an optimization to be a blocker for this PR.

```suggestion
    let mut instructions = script.instructions();

    let pubkey = match instructions.nth(PUBKEY_INDEX) {
    ...
    };

    let stake_hash = match instructions.nth(STAKE_HASH_INDEX) {
    ...
    };
```